lib/web/conn_upgrade.go

```diff
@@ -176,3 +231,81 @@ func (conn *waitConn) Close() error {
 	conn.cancel()
 	return trace.Wrap(err)
 }
+
+type websocketALPNServerConn struct {
+	*websocket.Conn
+	readBuffer []byte
+	readError  error
+	writeMutex sync.Mutex
+}
+
+func newWebSocketALPNServerConn(wsConn *websocket.Conn) *websocketALPNServerConn {
+	return &websocketALPNServerConn{
+		Conn: wsConn,
+	}
+}
+
+func (c *websocketALPNServerConn) convertError(err error) error {
+	if isOKWebsocketCloseError(err) {
+		return io.EOF
+	}
+	return err
+}
+
+func (c *websocketALPNServerConn) Read(b []byte) (n int, err error) {
```

Comment:
Should there be a mutex for reading as well ? My fear is concurrent read/write on `readBuffer` but is there some convention that .Read is not safe to call concurrently on a Conn ?

Reply:
added